### PR TITLE
patch caBundle for webhook config in remote cluster

### PR DESCRIPTION
**Pull request by irisdingbj**
https://github.com/istio/istio/issues/23561

### CODE REVIEW

File: manifests/charts/base/templates/clusterrole.yaml
Comment (howardjohn) [inline manifests/charts/base/templates/clusterrole.yaml:134]: Note - this role is called `istio-reader` but it essentially gives root control. This seems a bit surprising. It is blocked by `centralIstiod` though so maybe not a huge issue

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:144]: In the single cluster mode `injectionWebhookConfigName` is an environment variable. Seems like maybe we should use that as well, otherwise multi-revision will not work

File: pkg/util/webhookpatch.go
Comment (howardjohn) [inline pkg/util/webhookpatch.go:179]: why ignore the error here?

**Comment by howardjohn** [comment]
Overall looks good. Thanks!

File: manifests/charts/base/templates/clusterrole.yaml
Comment (linsun) [reply manifests/charts/base/templates/clusterrole.yaml:134]: yes, there is a https://github.com/istio/istio/issues/23263 to clean this

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (linsun) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:144]: should we guard this with centralIstiod before patch?

Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:144]: m.fetchCaRoot will not be nil only if we are on centralIstiod, so this has been garded.

Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:144]: injectionWebhookConfigName is defined in pilot/pkg/bootstrap/sidecarinjector.go. I have not found a proper way to reuse it in pilot/pkg/serviceregistry/kube/controller/multicluster.go. Any suggestions for this?

File: pkg/util/webhookpatch.go
Comment (irisdingbj) [reply pkg/util/webhookpatch.go]: log the error now

**Comment by irisdingbj** [comment]
@linsun @howardjohn Thanks for the review, PTAL.

**Comment by linsun** [comment]
LGTM, pending some successful manual test.

@hzxuzhonghu can you take a look pls and approve?

**Comment by irisdingbj** [comment]
manually test this out. caBundle in both mutatingwebhookconfiguration and validatingwebhookconfiguration have been patched

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/multicluster.go:145]: We'd better not do this in the utils, which should include only functions.

Same applies to `PatchCertLoop`.

I think they maybe ` pkg/webhooks` is a good place for them

File: pkg/util/webhookpatch.go
Comment (hzxuzhonghu) [inline pkg/util/webhookpatch.go:161]: We can unify the mutating/validating webhookconfiguration reconcile in a separate pr. Now one is patch and the other is update

File: pilot/pkg/bootstrap/server.go
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/server.go:229]: Can you point me why we need this field?

Comment (irisdingbj) [reply pilot/pkg/bootstrap/server.go:229]: we will use it in pilot/pkg/serviceregistry/kube/controller/multicluster.go:L144 :L145 to patch the webhook's caBundle

File: pkg/util/webhookpatch.go
Comment (irisdingbj) [reply pkg/util/webhookpatch.go:161]: this is to reuse original code that validationwebhook is using `update` while mutationwebhook is using `patch`

File: pilot/pkg/serviceregistry/kube/controller/multicluster.go
Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:145]: agree, think we can address this is a sperate PR?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:145]: I think this is to be backported to 1.6, right? If so, please address it in this one?

Comment (irisdingbj) [reply pilot/pkg/serviceregistry/kube/controller/multicluster.go:145]: to get into 1.6, need to be part of master first :)

**Review by hzxuzhonghu — APPROVED** [review]
As code freeze is close, i will approve first. Please do the cleanup in the following.

File: pilot/pkg/bootstrap/server.go
Comment (linsun) [reply pilot/pkg/bootstrap/server.go:229]: @hzxuzhonghu caBundle is required for remote webhook configuration to talk to the webhook server.

**Review by linsun — APPROVED** [review]
Note: this feature is not enabled by default.

**Comment by irisdingbj** [comment]
@douglas-reid @mandarjog Please take a look for this PR especially the mixer part. Thanks a lot!

**Review by douglas-reid — APPROVED** [review]

**Pull request merged**
